### PR TITLE
Fix for issue #5182: Allow cancel editing group infobox

### PR DESCRIPTION
**Pull request by vinsonizer**
Added cancel button that is available only when editing a group, which reverts changes from the server when clicked.  Also hides the "leave" button when in "Edit Mode"

### CODE REVIEW

File: website/public/js/controllers/groupsCtrl.js [hunk 1]
```diff
@@ -47,6 +47,11 @@ habitrpg.controller("GroupsCtrl", ['$scope', '$rootScope', 'Shared', 'Groups', '
       group._editing = false;
     };
 
+    $scope.cancelSave = function(group) {
+      group = group.$get();
```
Comment (crookedneighbor) [inline website/public/js/controllers/groupsCtrl.js]: Does the `$get` method return the group object? I thought `group.$get` returned a promise.

Comment (vinsonizer) [inline website/public/js/controllers/groupsCtrl.js]: According to the resource docs (
https://docs.angularjs.org/api/ngResource/service/$resource) it returns an
empty object that is populated on result which will trigger rendering
changes. I guess this could create some interesting client experience, but
I wasn't sure if going with something like a shadow copy directive would
over complicate the change.

Also I am new to Angular and Jade, so there is likely an even better way to
handle this.

**Comment by vinsonizer** [comment]
User ID b57a198e-d079-48ef-9fd9-631d99a628ce by the way

File: website/public/js/controllers/groupsCtrl.js (hunk 1 above)
Comment (crookedneighbor) [inline website/public/js/controllers/groupsCtrl.js]: You shouldn't need to reassign the variable, just doing:

``` js
$scope.cancelSave = function(group) {
  group.$get();
  //...
```

Should work.

That said, I think making a local copy is probably preferable so we don't need to make an extra server call for the cancel activity to work. This will probably mean making an `$scope.editGroup` function.

Comment (crookedneighbor) [inline website/public/js/controllers/groupsCtrl.js]: That said, I'm not an expert on client side performance.

Also, thanks for the link. I really need to actually read the docs for ng-resource. :smile:

**Comment by crookedneighbor** [comment]
In addition, we'll need to add tests for the new functions. [the group controller doesn't have a ton of coverage right now](https://github.com/HabitRPG/habitrpg/blob/develop/test/spec/controllers/groupCtrlSpec.js), but we should still add tests for new functionality.

**Comment by vinsonizer** [comment]
I am not familiar with the testing framework used, but I will see if i can get something together that will work.

**Comment by crookedneighbor** [comment]
If you need help, I'm happy to give it.

**Comment by vinsonizer** [comment]
Just need a start, glad to do it if you can point me to references

**Comment by vinsonizer** [comment]
@crookedneighbor I have written some test cases for the groups controller, but i am having issues actually executing the save method (undefined).  Is there some way i can share here without committing into the pull request?

**Comment by vinsonizer** [comment]
Actually, you can see it in this gist, from line 74 on:

https://gist.github.com/vinsonizer/55d5a8cc444d77ccba85

I get errors on the last line of the test, which is the invocation of group.$save() in the controller.

**Comment by crookedneighbor** [comment]
@vinsonizer looks like maybe you rebased and pulled in a merge commit? Could you remove that extraneous commit from this PR so it's easier to review?

**Comment by vinsonizer** [comment]
@crookedneighbor how do I do that?  I've rebased twice trying to ensure i stay up to date, but I am getting duplicate merge conflicts.  I am not sure why that is...

**Comment by vinsonizer** [comment]
@crookedneighbor Cleaning up now, hopefully have it all recommitted today

**Comment by vinsonizer** [comment]
@crookedneighbor  Ok I reset (further back than I meant to) and reapplied my changes.  Please let me know if this is what you are looking for.

**Comment by vinsonizer** [comment]
@crookedneighbor @SabreCat I hate to keep hitting you guys with new commits... I now understand the jasmine library and am restructuring the group controller tests (all of them) into more a modular setup.  Please wait to review the changes till I've completed this.  I think it will be much nicer.

**Comment by crookedneighbor** [comment]
@vinsonizer No worries. Ping me when you're ready for me to look at it.

**Comment by vinsonizer** [comment]
@crookedneighbor I think it is done.  There are only two sections I didn't add tests for: 1.) Modal methods and 2.) isMember, which appears to be using coercion and also basically looks like it should be superseded by isMemberOfGroup.

File: test/spec/controllers/groupCtrlSpec.js [hunk 2]
```diff
@@ -23,51 +23,207 @@ describe('Groups Controller', function() {
...
+      expect(guild.description).to.eql(editGuild.description);
+      expect(guild._newLeader).to.eql(editGuild._newLeader);
+      expect(guildSave).to.be.calledOnce;
 
-      expect(scope.isMemberOfGroup(user._id, guild)).to.not.be.ok;
-      expect(myGuilds).to.be.calledOnce;
     });
   });
+  /* TODO: Modal testing
+  describe("deleteAllMessages", function() { });
```
Comment (crookedneighbor) [inline test/spec/controllers/groupCtrlSpec.js]: The proper way to pend these is by doing:

``` js
describe.skip('deleteAllMessages', function () {

});
```

File: test/spec/controllers/groupCtrlSpec.js [hunk 3]
```diff
@@ -23,51 +23,207 @@ describe('Groups Controller', function() {
...
+    it('should return false if my user isn\'t part of the quest members', function() {
+      pendingQuestParty.quest.members = {};
+      runningQuestParty.quest.members = {};
+      expect(scope.isMemberOfPendingQuest(user._id, pendingQuestParty)).to.eql(false);
+      expect(scope.isMemberOfRunningQuest(user._id, runningQuestParty)).to.eql(false);
+    });
+  });
+   
+
+  describe("Group Edit Behavior", function() {
```
Comment (crookedneighbor) [inline test/spec/controllers/groupCtrlSpec.js]: This is describing the `editGroup` method, so it should be `describe('editGroup'`, function () {`

File: test/spec/controllers/groupCtrlSpec.js [hunk 4]
```diff
@@ -23,51 +23,207 @@ describe('Groups Controller', function() {
...
+
```
Comment (crookedneighbor) [inline test/spec/controllers/groupCtrlSpec.js]: No blank line here.

File: test/spec/controllers/groupCtrlSpec.js [hunk 5]
```diff
@@ -23,51 +23,207 @@ describe('Groups Controller', function() {
...
+    it('should return false if my user isn\'t part of the quest members', function() {
+      pendingQuestParty.quest.members = {};
+      runningQuestParty.quest.members = {};
+      expect(scope.isMemberOfPendingQuest(user._id, pendingQuestParty)).to.eql(false);
+      expect(scope.isMemberOfRunningQuest(user._id, runningQuestParty)).to.eql(false);
+    });
+  });
+   
+
```
Comment (crookedneighbor) [inline test/spec/controllers/groupCtrlSpec.js]: There's an extra line break here.

File: test/spec/controllers/groupCtrlSpec.js [hunk 6]
```diff
@@ -23,51 +23,207 @@ describe('Groups Controller', function() {
...
+      var editGuild = scope.groupCopy;
+      expect(editGuild).to.eql({});
+      expect(guild._editing).to.eql(undefined);
+    });
+
+
+    it('should allow for editing without changing group resource', function() {
+      // switch to edit mode
+      scope.editGroup(guild);
+      var editGuild = scope.groupCopy;
```
Comment (crookedneighbor) [inline test/spec/controllers/groupCtrlSpec.js]: Why is it important to assign this to a variable? Why can't you assert on `scope.groupCopy` directly?

File: website/public/js/controllers/groupsCtrl.js [hunk 7]
```diff
@@ -12,7 +12,7 @@ habitrpg.controller("GroupsCtrl", ['$scope', '$rootScope', 'Shared', 'Groups', '
     $scope.isMemberOfRunningQuest = function(userid, group) {
       if (!group.quest || !group.quest.members) return false;
       if (!group.quest.active) return false; // quest is pending, not started
-      return group.quest.members[userid];
+      return userid in group.quest.members && group.quest.members[userid] != false;
```
Comment (crookedneighbor) [inline website/public/js/controllers/groupsCtrl.js:15]: Why are you working on this code? Seems like this should be a separate PR.

**Comment by crookedneighbor** [comment]
The group leader message box is not part of this PR: https://github.com/HabitRPG/habitrpg/pull/6400/files#diff-093f1d61737be6ef20f580e7dcc66fe1R134

File: test/spec/controllers/groupCtrlSpec.js (hunk 4 above)
Comment (vinsonizer) [inline test/spec/controllers/groupCtrlSpec.js]: Which line specifically?  149?

File: website/public/js/controllers/groupsCtrl.js (hunk 7 above)
Comment (vinsonizer) [inline website/public/js/controllers/groupsCtrl.js:15]: When adding tests, i found that the behavior of this method and isMemberOfPendingQuest was inconsistent.  This should be functionally equivalent to what was there before but also handles undefined.  It is a safer check overall right?

In reality these two methods should probably be refactored for code reuse, since they are almost identical other than the active check.

**Comment by vinsonizer** [comment]
@crookedneighbor 

I will add the leader message box in later this evening and incorporate your comments.  If you can give specific feedback on the whitespace concerns, i will also correct those.

I already fixed the other comments (re: describe.skip, describe editGroup, local variable assignment)

**Comment by crookedneighbor** [comment]
I've merged in the portion of your pr that deals with the editGroup. If you'd like to open a separate PR with your isMember changes, feel free.

Note: I changed your test to be in a more unit test style instead of a behavioral style: https://github.com/HabitRPG/habitrpg/commit/2989057dd3093247472010acb24d9b88b0905ee2

Tier awarded! You can read more about [contributor tiers and rewards here](http://habitrpg.wikia.com/wiki/Contributor_Rewards).

Note that as your tier increases, it takes progressively more work to achieve the next one, so for the higher tiers, larger PRs or a few PRs are required. There's some information about that at Contributor Rewards, but feel free to ask here if you have questions!
